mlir/lib/Conversion/TosaToRock/TosaToRock.cpp

```diff
@@ -398,39 +398,112 @@ struct TransposeRewritePattern : public OpRewritePattern<tosa::TransposeOp> {
     op->setAttr(attrKey, StringAttr::get(op->getContext(), layout));
   }
 
-  void setTranspose(Operation *op, StringRef name, bool isNonTrivial) const {
+  void setTranspose(Operation *op, StringRef name, bool isTrivial) const {
+    bool isNonTrivial = !isTrivial;
     bool currentValue = false;
     if (auto attr = op->getAttrOfType<BoolAttr>(name))
       currentValue = attr.getValue();
     bool newValue = currentValue ^ isNonTrivial;
     op->setAttr(name, BoolAttr::get(op->getContext(), newValue));
   }
 
+  LogicalResult checkMatMulTransposeValid(tosa::MatMulOp matmulOp,
+                                          ArrayRef<int32_t> dims) const {
+    // batch dimension is expected to be 3rd from the last.
+    if (dims.size() >= 3 && dims[dims.size() - 3] != (int32_t)dims.size() - 3) {
+      return matmulOp.emitError(
+          "Can't transpose the batch dimension out of place");
+    }
+    return success();
+  }
+
+  bool isMatMulTrivial(ArrayRef<int32_t> dims) const {
+    return (dims[dims.size() - 2] == 2 && dims[dims.size() - 1] == 1);
+  }
+
+  LogicalResult mergeTransposeWithGemmLikeOp(PatternRewriter &rewriter,
+                                             Value tOutput,
+                                             ArrayRef<int32_t> dims,
+                                             Value &tInput) const {
+    for (auto &use : tOutput.getUses()) {
+      if (auto op = dyn_cast<tensor::CollapseShapeOp>(use.getOwner())) {
+        SmallVector<ReassociationIndices, 4> reassocIndices =
+            op.getReassociationIndices();
+        tensor::CollapseShapeOp newCollapseShapeOp =
+            rewriter.create<tensor::CollapseShapeOp>(op.getLoc(), tInput,
+                                                     reassocIndices);
+        tInput = newCollapseShapeOp.getResult();
+        if (mergeTransposeWithGemmLikeOp(rewriter, op.getResult(), dims, tInput)
+                .failed()) {
+          return failure();
+        }
+        rewriter.eraseOp(op);
+      } else if (auto op = dyn_cast<tensor::ExpandShapeOp>(use.getOwner())) {
+        SmallVector<ReassociationIndices, 4> reassocIndices =
+            op.getReassociationIndices();
+        ArrayRef<int64_t> outShape = op.getResultType().getShape();
+        ArrayRef<int64_t> inShape = op.getSrcType().getShape();
+        SmallVector<int64_t, 4> newOutShape = llvm::to_vector<4>(outShape);
+        // Assumption is shape expansions happen only in higher dimensions.
+        for (size_t i = 0; i < dims.size(); i++) {
+          newOutShape[newOutShape.size() - 1 - i] =
```

Comment:
Basically, I want to lift the expand shape on top of transpose.

I thought about faithfully following the reassociation expressions but then it would associate expanded dim strictly as before -- which is oddly what I dont want.

E.g. 

Before:
tp:                    m x n ---> n x m      
expand_shape: n x m ---> 1 x n x m
                         reassoc : [0, 1] , [2]

After (what I want it to be with no tp)
new_expand_shape: m x n --> 1 x m x n
                                 reassoc : [0, 1] , [2]  is still the same.
                                 
So reassociation seems identical -- just some dims are swapped. 
Here I just wanted to calculate the outShape of new_expand_shape while using the old reassociation.

I maybe wrong -- but strictly associating with reassociation felt like it would be give something like m x 1 x n which is not what we want here ?

Reply:
or I could drop expand shape support -- we dont strictly need it (yet). What I m seeing is expanded shapes coming from clients (that has more dims than the ops require) and then our lowering add reshapes to collapse.

Reply:
removed it for now -- it seems confusing and I dont see us exercising the case either (unless in a contrived test) -- we can revisit if needed.

Reply:
... ah, that makes more sense. 

I think the cases where we have expand_shape after this all could be stuff like adding a group dimension ... but that's usually done with coordinate transforms.